sdks/unity/AgonesSdk.cs

```diff
@@ -0,0 +1,239 @@
+        {
+            string json = JsonUtility.ToJson(new KeyValueMessage(key, value));
+            return await SendRequestAsync("/metadata/annotation", json, UnityWebRequest.kHttpVerbPUT);
+        }
+        #endregion
+
+        #region AgonesRestClient Private Methods
+        void Health()
+        {
+            _ = SendRequestAsync("/health", "{}");
+        }
+
+        async Task<bool> SendRequestAsync(string api, string json, string method = UnityWebRequest.kHttpVerbPOST)
+        {
+            // To prevent that an async method leaks after destroying this gameObject
+            cancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+            var req = new UnityWebRequest(sidecarAddress + api, method)
+            {
+                uploadHandler = new UploadHandlerRaw(Encoding.UTF8.GetBytes(json)),
+                downloadHandler = new DownloadHandlerBuffer()
+            };
+            req.SetRequestHeader("Content-Type", "application/json");
+
+            await new AgonesAsyncOperationWrapper(req.SendWebRequest());
+
+            bool ok = req.responseCode == 200;
+
+            if (ok)
+            {
+                Log($"Agones SendRequest ok: {api} {req.downloadHandler.text}");
+            }
+            else
+            {
+                Log($"Agones SendRequest failed: {api} {req.error}");
+            }
+
+            return ok;
+        }
+
+        void Log(object message)
+        {
+            if (!logEnabled) { return; }
+
+#if UNITY_EDITOR
+            Debug.Log(message);
+#else
+            Console.WriteLine(message);
+#endif
+        }
+        #endregion
+
+        #region AgonesRestClient Nested Classes
+        class AgonesAsyncOperationWrapper
+        {
+            public UnityWebRequestAsyncOperation AsyncOp { get; }
+            public AgonesAsyncOperationWrapper(UnityWebRequestAsyncOperation unityOp)
+            {
+                AsyncOp = unityOp;
+            }
+
+            public AgonesAsyncOperationAwaiter GetAwaiter()
+            {
+                return new AgonesAsyncOperationAwaiter(this);
+            }
+        }
+
+        class AgonesAsyncOperationAwaiter : INotifyCompletion
+        {
+            private UnityWebRequestAsyncOperation asyncOp;
+            private Action continuation;
+
+            public AgonesAsyncOperationAwaiter(AgonesAsyncOperationWrapper wrapper)
+            {
+                asyncOp = wrapper.AsyncOp;
+                asyncOp.completed += OnRequestCompleted;
+            }
+
+            public bool IsCompleted => asyncOp.isDone;
+
+            public void GetResult()
+            {
+                asyncOp.completed -= OnRequestCompleted;
+
+                // remove references
+                asyncOp = null;
+                continuation = null;
+            }
+
+            public void OnCompleted(Action continuation)
+            {
+                this.continuation = continuation;

```

Comment:
This class doesn't appear to work if the async operation has already completed by the time this Awaiter is created.

Unless this is impossible, this function should check if the operation is complete and if so, run the callback immediately.

And whenever the continuation function is called, the `continuation` field should be nulled out so as not to be repeated. 

Please also include some sort of `lock` statement on the `continuation` field itself to prevent race conditions.

Reply:
I can not explain in English well, so I paste the IL code of `SendRequestAsync` that is decompiled with [dnSpy](https://github.com/0xd4d/dnSpy).
```csharp
void IAsyncStateMachine.MoveNext()
{
    int num = this.<>1__state;
    AgonesSdk agonesSdk = this.<>4__this;
    bool result;
    try
    {
        AgonesSdk.AgonesAsyncOperationAwaiter agonesAsyncOperationAwaiter;
        if (num != 0)
        {
            agonesSdk.cancellationTokenSource.Token.ThrowIfCancellationRequested();
            this.<req>5__2 = new UnityWebRequest("http://localhost:59358" + this.api, this.method)
            {
                uploadHandler = new UploadHandlerRaw(Encoding.UTF8.GetBytes(this.json)),
                downloadHandler = new DownloadHandlerBuffer()
            };
            this.<req>5__2.SetRequestHeader("Content-Type", "application/json");
            agonesAsyncOperationAwaiter = new AgonesSdk.AgonesAsyncOperationWrapper(this.<req>5__2.SendWebRequest()).GetAwaiter();
            if (!agonesAsyncOperationAwaiter.IsCompleted)
            {
                this.<>1__state = 0;
                this.<>u__1 = agonesAsyncOperationAwaiter;
                // this methods means like OnCompleted(()=>{ this.MoveNext(); })
                this.<>t__builder.AwaitOnCompleted<AgonesSdk.AgonesAsyncOperationAwaiter, AgonesSdk.<SendRequestAsync>d__14>(ref agonesAsyncOperationAwaiter, ref this);
                return;
            }
        }
        else
        {
            agonesAsyncOperationAwaiter = (AgonesSdk.AgonesAsyncOperationAwaiter)this.<>u__1;
            this.<>u__1 = null;
            this.<>1__state = -1;
        }
        agonesAsyncOperationAwaiter.GetResult();
        bool flag = this.<req>5__2.responseCode == 200L;
        if (flag)
        {
            agonesSdk.Log("Agones SendRequest ok: " + this.api + " " + this.<req>5__2.downloadHandler.text);
        }
        else
        {
            agonesSdk.Log("Agones SendRequest failed: " + this.api + " " + this.<req>5__2.error);
        }
        result = flag;
    }
    catch (Exception exception)
    {
        this.<>1__state = -2;
        this.<>t__builder.SetException(exception);
        return;
    }
    this.<>1__state = -2;
    this.<>t__builder.SetResult(result);
}
```

I think that if the async operation completed immediately, this code works well. 
And `OnCompleted` is called on the main thread, so the `lock` statement is not necessary.

Please check it.